pkg/util/podstore/podstore.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package podstore
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
+)
+
+const (
+	DeleteQueueName = "pod_delete"
+	PodIPIndex      = "PodIP"
+	DelayTime       = time.Minute * 5
+)
+
+type PodStore struct {
+	curPods      cache.Indexer
+	prevPods     cache.Indexer
+	podsToDelete workqueue.DelayingInterface
+	// Mapping pod.uuid to PodLocalTimestamps
+	timestampMap map[string]*podLocalTimestamps
+	clock        clock.Clock
+	mutex        sync.Mutex
+}
+
+type podLocalTimestamps struct {
+	// The local clock when the Pod is added
+	CreationTimestamp time.Time
+	// The local clock when the Pod is actually deleted from kube-apiserver. Nil if it's not deleted yet.
+	DeletionTimestamp *time.Time
+}
+
+// NewPodStoreWithClock creates a Pod Store with a custom clock,
+// which is useful when writing robust unit tests.
+func NewPodStoreWithClock(podInformer cache.SharedIndexInformer, clock clock.WithTicker) *PodStore {
+	s := &PodStore{
+		curPods:      cache.NewIndexer(podKeyFunc, cache.Indexers{PodIPIndex: podIPIndexFunc}),
+		prevPods:     cache.NewIndexer(podKeyFunc, cache.Indexers{PodIPIndex: podIPIndexFunc}),
+		podsToDelete: workqueue.NewDelayingQueueWithCustomClock(clock, DeleteQueueName),
+		clock:        clock,
+		timestampMap: map[string]*podLocalTimestamps{},
+	}
+	podInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    s.onPodCreate,
+			DeleteFunc: s.onPodDelete,
+		})
+	return s
+}
+
+func NewPodStore(podInformer cache.SharedIndexInformer) *PodStore {
+	return NewPodStoreWithClock(podInformer, clock.RealClock{})
+}
+
+func (s *PodStore) onPodCreate(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		klog.ErrorS(nil, "Received unexpected object", "obj", obj)
+		return
+	}
+	err := s.curPods.Add(pod)
+	if err != nil {
+		klog.ErrorS(err, "Error when adding pod to curPods")
+		return
+	}
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	timeNow := s.clock.Now()
```

Comment:
One thing that' still not ideal is that when the PodStore is used in the FlowAggregator, the clock of that Node may not match the clock on the Node running FlowExporter (which records the connection start time).

Maybe we could slightly improve the code in `GetPodByIPAndTime` as follows: if we cannot find a Pod matching both IP + time, but there is a single Pod matching just the IP, we return that Pod. What do you think? Any drawback to this approach?

Reply:
Will the difference of clock be huge between two different nodes? But even the clock might be different, is it possible that a traffic starts before PodStores receive the Add event in FE and FA?

The only concern is that, if we have two Pod (one in curPods, one in prevPods) with same IP in our FA. Which do one we want to return?

Suppose, we have PodA (IP=1.1.1.1, CreationTime=14) in FE and PodA (IP=1.1.1.1, CreationTime=14.5) in FA.
And we also have PodB (IP=1.1.1.1, CreationTime=12, DeletionTime=14) in prevPods in FA.
If we have a connection with startTime=14.2, we can find the correct pod in FE. But we can't find the right Pod in FA?

Reply:
I'm also thinking that if FA restarts, the creationTime for curPods will be wrong, right?
And what's the case if a Pod is created before we start the PodInformer in FA? Will we still receive the Add event?

Reply:
> is it possible that a traffic starts before PodStores receive the Add event in FE and FA?

yes it's possible. I don't know if you want to do anything about that? This is not an issue at the FA because of the larger delay. It may be an issue at the FE. Here there is no choice but waiting or querying the interface store.

> The only concern is that, if we have two Pod (one in curPods, one in prevPods) with same IP in our FA. Which do one we want to return?

We don't return anything in that case. There is more than one Pod matching just the IP.

> I'm also thinking that if FA restarts, the creationTime for all pods in curPods will be wrong, right?

Yes, that's a bit of a concern. But as long as we do not return incorrect information, it's ok. Returning empty / missing is better than returning erroneous values.

Reply:
> It may be an issue at the FE. Here there is no choice but waiting or querying the interface store.
I think it will be an issue at FE/FA. Especially when we have FE/FA restart.
Since the Pod's CreationTime is later than the connection startTime, we won't be able to fill the pod information/labels in FE/FA . And the Pod Info will be missing forever for this connection.

Is it possible that the Pod's default CreationTime (set by the K8s apiserver) is later than the connection startTime?

Although we were saying this default CreationTime may be skewed compared to the local one, I found it is pretty close to the local clock. The time difference is caused by the unit they used (CreationTime is in second, but our clock is in millisecond.) By using the default CreationTime, we can still get the correct Pod's CreationTime even when FE/FA restarts.

What do you think? Or we should create a special case in `GetPodByIPAndTime`?

Reply:
> Although we were saying this default CreationTime may be skewed compared to the local one, I found it is pretty close to the local clock.

This will depend on the cluster. It is is possible to have one or more Node with an "incorrect" clock. However, we could claim that it is the cluster admin's responsibility to avoid that. Some K8s managed services provide time synchronization.

Given everything we have discussed, it seems reasonable to:
- revert back to CreationTimestamp for the Pod "created" time
- stick to "local wall-clock time" for Pod "deleted" time

I would still recommend implementing the suggestion above:
> if we cannot find a Pod matching both IP + time, but there is a single Pod matching just the IP, we return that Pod.